### PR TITLE
Reapply "Newton error propagation based convergence check."

### PR DESCRIPTION
**Pull request by 1uc**
Second attempt at implementing Newton error propagation
based convergence check.

Adds `EIGEN_DEVICE_FUNC` to `is_converged` for GPU builds.

### CODE REVIEW

**Comment by 1uc** [comment]
CI went green for NRN with this PR: https://github.com/neuronsimulator/nrn/pull/3068

File: src/solver/newton/newton.hpp [hunk 1]
```diff
@@ -34,8 +34,33 @@ namespace newton {
  * @{
  */
 
-static constexpr int MAX_ITER = 1e3;
-static constexpr double EPS = 1e-12;
+static constexpr int MAX_ITER = 50;
```
Comment (pramodk) [inline src/solver/newton/newton.hpp:37]: Don't have relevant expertise and hence curious about which aspects allow much lower max iterations.

**Review by pramodk — APPROVED** [review]
LGTM

File: src/solver/newton/newton.hpp (hunk 1 above)
Comment (1uc) [reply src/solver/newton/newton.hpp:37]: NEURON uses 20.

If Newton converges, the number of `0`s in the error doubles with each iteration, i.e. from 0.1 error it takes `3` iterations to hit `1e-14`. Unlike stochastic gradient descent, if it doesn't converge, there's no nothing to make it eventually converge.

**Pull request merged**